weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -1413,6 +1405,144 @@ def _insert_call_batch(self, batch: list) -> None:
                 settings=settings,
             )
 
+    def _select_single_obj_query(
+        self,
+        project_id: str,
+        object_id: str,
+        conditions: list[str],
+    ) -> SelectableCHObjSchema:
+        """Query for reading a single object by digest.
+
+        First, query all versions of the object by id to calculate the version index,
+        without the overhead of val_dump.
+        Then, run a second query to get the complete object with the given digest by
+        directly filtering on the digest.
+        """
+        condition_part = combine_conditions(conditions, "AND")
+
+        # Get version counting fields
+        version_index_and_count_query = f"""
+            SELECT
+                digest,
+                version_index,
+                version_count,
+                is_latest
+            FROM (
+                SELECT
+                    digest,
+                    row_number() OVER (
+                        PARTITION BY project_id,
+                        kind,
+                        object_id
+                        ORDER BY created_at ASC
+                    ) - 1 AS version_index,
+                    count(*) OVER (PARTITION BY project_id, kind, object_id) as version_count,
+                    if(version_index + 1 = version_count, 1, 0) AS is_latest
+                FROM (
+                    SELECT
+                        digest,
+                        project_id,
+                        kind,
+                        object_id,
+                        created_at,
+                        row_number() OVER (
+                            PARTITION BY project_id,
+                            kind,
+                            object_id,
+                            digest
+                            ORDER BY created_at ASC
+                        ) AS rn
+                    FROM object_versions
+                    WHERE project_id = {{project_id: String}} AND
+                        object_id = {{object_id: String}}
+                )
+                WHERE rn = 1
+            )
+            WHERE {condition_part}
+        """
+        query_result = self._query(
+            version_index_and_count_query,
+            {"project_id": project_id, "object_id": object_id},
+        )
+        if len(query_result.result_rows) == 0:
+            raise NotFoundError(f"Obj {object_id} not found")
+
+        (digest, version_index, version_count, is_latest) = query_result.result_rows[0]
+
+        # Now get the full object by digest
+        select_query = """
+            SELECT
+                project_id,
+                object_id,
+                created_at,
+                kind,
+                base_object_class,
+                refs,
+                val_dump,
+                if (kind = 'op', 1, 0) AS is_op
+            FROM (
```

Comment:
If you extract everything except `val_dump` from the first query, then this query can avoid the 2-layer query by ignoring row_number stuff

Reply:
Like we allow for re-inserting objects, so there could be more than one with the same primary key (briefly). Re that issue that the data team mentioned wrt to changing `created_at` times.

Reply:
although in this case val_dump would be the same... Okay i see

Reply:
The `created_at` could be different, but we probably don't care

Reply:
get created at from the first one

Reply:
yessir, check it now